src/solver/algorithm/strong/puzzle.rs

```diff
@@ -9,95 +9,233 @@ use anyhow::{Context, Result};
+        let parents = game.retrograde(state);
+        let child_remoteness =
+            RecordBuffer::from(db.get(state).unwrap())?.get_remoteness();
+
+        for parent in parents {
+            let child_count =
+                RecordBuffer::from(db.get(parent).unwrap())?.get_child_count();
+            if child_count > 0 {
+                // Update child count
+                let mut buf = RecordBuffer::from(db.get(parent).unwrap())
+                    .context("Failed to get record for middle state")?;
+                let new_child_count = buf.get_child_count() - 1;
+                buf.set_child_count(new_child_count)?;
+                db.put(parent, &buf);
+
+                // If all children have been solved, set this state as a losing
+                // state
+                if new_child_count == 0 {
+                    losing_queue.push_back(parent);
+                    update_db_record(
+                        db,
+                        parent,
+                        SimpleUtility::LOSE,
+                        1 + child_remoteness,
+                        0,
+                    )?;
+                }
+            }
+        }
+    }
+
+    Ok(())
 }
 
-fn bfs_state<G, D>(db: &mut D, game: &G)
+/// Updates the database record for a puzzle with given simple utility,
+/// remoteness, and child count
+fn update_db_record<D>(
+    db: &mut D,
+    state: State,
+    utility: SimpleUtility,
+    remoteness: Remoteness,
+    child_count: ChildCount,
+) -> Result<()>
 where
-    G: DTransition<State> + Bounded<State> + SimpleSum<N>,
-    D: KVStore<RecordBuffer>,
+    D: KVStore,
 {
+    let mut buf = RecordBuffer::from(db.get(state).unwrap())
+        .context("Failed to create record for middle state")?;
+    buf.set_utility([utility])
+        .context("Failed to set utility for state.")?;
+    buf.set_remoteness(remoteness)
+        .context("Failed to set remoteness for state.")?;
+    buf.set_child_count(child_count)
+        .context("Failed to set child count for state.")?;
+    db.put(state, &buf);
 
+    Ok(())
 }
 
-fn discover_end_states<G, D>(db: &mut D, game: &G) -> Vec<State> 
+fn discover_child_counts<G, D>(db: &mut D, game: &G) -> Result<Vec<State>>
 where
-    G: DTransition<State> + Bounded<State> + SimpleSum<N>,
-    D: KVStore<RecordBuffer>,
+    G: DTransition<State>
+        + Bounded<State>
+        + ClassicPuzzle
+        + Extensive<1>
+        + Game,
+    D: KVStore,
 {
-    let visited = HashSet::new();
-    let end_states = Vec::new();
-
-    discover_end_states(db, game, game.start(), visited, end_states);
+    let mut end_states = Vec::new();
+    discover_child_counts_helper(db, game, game.start(), &mut end_states)?;
 
-    end_states
+    Ok(end_states)
 }
 
-fn discover_end_states_helper<G, D>(db: &mut D, game: &G, state: State, visited: HashSet<State>, end_states: Vec<State>)
+/// Adds child counts for each position to the database
```

Comment:
If you add a docstring, it should be grammatically correct / well formatted (they end up in the project documentation, which looks super wonky when they're not properly written).

Alternatively, just don't add a docstring (only applies to private items, so anything that is not `pub`).